#! /usr/bin/env bash

#BSUB -J peaks[1-13]
#BSUB -e peaks.%J.%I.err
#BSUB -o peaks.%J.%I.out
#BSUB -q short

<<DOC
call peaks using macs2.
DOC

set -o nounset -o pipefail -o errexit -x

source $CONFIG
sample=${SAMPLES[$(($LSB_JOBINDEX - 1))]}

results=$RESULT/$sample
peakresults="$results/peaks"

# yeast genome size
genomesize=12e6

# broadPeak autosql
ucscdir=/vol1/software/modules-sw/ucsc/build/v286
asfile=$ucscdir/kent/src/hg/lib/encode/broadPeak.as

# define extsize here to filter for peaks of size 1 later
extsize=25

if [[ ! -f $peakresults ]]; then
    mkdir -p $peakresults
fi

strands=("all" "pos" "neg")

for strand in ${strands[@]}; do
    for align_mode in ${ALIGN_MODES[@]}; do

        exp_name="$peakresults/$sample.$strand.align.$align_mode"

        # these filenames are generated by macs2
        peak="${exp_name}_peaks.bed"
        broadpeak="${exp_name}_peaks.broadPeak"

        bigbed="${exp_name}_peaks.bb"

        bam=$results/alignment/$sample.align.$align_mode.bam

        macs2 callpeak -t $bam \
            -n $exp_name \
            --keep-dup all \
            --nomodel \
            --gsize $genomesize \
            --extsize $extsize \
            --broad

        # sometimes the score exceeds the maximum (1000) defined by the
        # broadPeak spec. Reformat the broadPeak file to covert > 1000
        # to 1000
        broadpeak_tmpfile="$broadpeak.tmp"
        awk 'BEGIN {OFS="\t"} \
                   { if ($5 > 1000) $5 = 1000; print $0}' \
            < $broadpeak \
            | awk -v extsize=$extsize '$3 - $2 > extsize' \
            > $broadpeak_tmpfile
        mv $broadpeak_tmpfile $broadpeak
        
        bedToBigBed -type=bed6+3 -as=$asfile \
            $broadpeak $CHROM_SIZES $bigbed

    done
done

